### PR TITLE
Convert commands to Kotlin

### PR DESCRIPTION
**Pull request by cubeee**
Convert commands from Ruby to Kotlin with 1:1 functionality.

TODO:
- [x] Command registering
- [x] Commands:
    - [x]  animate
    - [x]  bank
    - [x]  item
    - [x]  lookup
    - [x]  messaging
    - [x]  npc
    - [x]  punishment
    - [x]  skill
    - [x]  teleport

### CODE REVIEW

File: game/src/main/java/org/apollo/game/plugin/KotlinPluginEnvironment.java [hunk 1]
```diff
@@ -1,30 +1,14 @@
 package org.apollo.game.plugin;
 
-import java.io.BufferedReader;
-import java.io.File;
-import java.io.InputStream;
-import java.io.InputStreamReader;
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
```
Comment (Major-) [inline game/src/main/java/org/apollo/game/plugin/KotlinPluginEnvironment.java]: Can you revert this change?

File: game/src/main/kotlin/org/apollo/game/plugin/kotlin/KotlinPluginScript.kt [hunk 2]
```diff
@@ -61,3 +66,22 @@ class KotlinMessageHandler<T : Message>(val world: World, val context: PluginCon
...
+        function.invoke(player, command)
+    }
+
+}
+
+class KotlinCommandHandler(val world : World, val command: String, val privileges: PrivilegeLevel) {
+
+    var function: (Player, Command) -> Unit = { _, _ -> }
+
+    fun then(function: (Player, Command) -> Unit) {
```
Comment (Major-) [inline game/src/main/kotlin/org/apollo/game/plugin/kotlin/KotlinPluginScript.kt]: If this is changed to `Command.(Player) -> Unit` then it'll be invoked from the `Command`'s context (just like `.then` in the message handlers) and you'll only have the lone `Player` argument.

File: game/src/main/java/org/apollo/game/plugin/KotlinPluginEnvironment.java (hunk 1 above)
Comment (cubeee) [inline game/src/main/java/org/apollo/game/plugin/KotlinPluginEnvironment.java]: Done, mb

File: game/src/main/kotlin/org/apollo/game/plugin/kotlin/KotlinPluginScript.kt (hunk 2 above)
Comment (cubeee) [inline game/src/main/kotlin/org/apollo/game/plugin/kotlin/KotlinPluginScript.kt]: Done, cheers

File: game/src/plugins/util/command/src/command.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import org.apollo.game.model.entity.Player
+
+/**
+ * Checks whether the amount of arguments provided is correct, sending the player the specified
+ * message if not.
+ */
+fun valid_arg_length(args: Array<String>, length: Any, player: Player, message: String): Boolean {
```
Comment (Major-) [inline game/src/plugins/util/command/src/command.kt]: I think its probably nicer to overload this method, providing one with `length: Int` and another with `length: IntRange`, to retain type safety

File: game/src/plugins/cmd/src/item-cmd.plugin.kts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import org.apollo.cache.def.ItemDefinition
+import org.apollo.game.model.entity.setting.PrivilegeLevel
+
+on_command("item", PrivilegeLevel.ADMINISTRATOR)
+        .then { player ->
+            if  (!valid_arg_length(arguments, 1..2, player, "Invalid syntax - ::item [id] [amount]")) {
+                return@then
```
Comment (Major-) [inline game/src/plugins/cmd/src/item-cmd.plugin.kts]: I think these can just be `return`?

File: game/src/plugins/cmd/src/broadcast-cmd.plugin.kts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import org.apollo.game.model.World
+import org.apollo.game.model.entity.setting.PrivilegeLevel
+
+var world: World? = null
```
Comment (Major-) [inline game/src/plugins/cmd/src/broadcast-cmd.plugin.kts]: @garyttierney this is pretty gross, anything we can do about this?

File: game/src/plugins/util/command/src/command.kt (hunk 3 above)
Comment (cubeee) [inline game/src/plugins/util/command/src/command.kt]: Done

File: game/src/plugins/cmd/src/item-cmd.plugin.kts (hunk 4 above)
Comment (cubeee) [inline game/src/plugins/cmd/src/item-cmd.plugin.kts]: 'return' is not allowed. Followed IDEA's autocomplete and that was the only option

Comment (Major-) [inline game/src/plugins/cmd/src/item-cmd.plugin.kts]: Ah apparently you need to label it unless the function is marked `inline`, my mistake. This is fine

File: game/src/plugins/cmd/src/skill-cmd.plugin.kts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ */
+on_command("xp", PrivilegeLevel.ADMINISTRATOR)
+        .then { player ->
+            val invalidSyntax = "Invalid syntax - ::xp [skill-id] [experience]"
+            if (arguments.size != 2) {
+                player.sendMessage(invalidSyntax)
+                return@then
+            }
+
+            val skillId = arguments[0].toInt()
```
Comment (Major-) [inline game/src/plugins/cmd/src/skill-cmd.plugin.kts]: These can throw `NumberFormatException` - maybe use Guava's `Ints/Doubles.tryParse` with a nullcheck instead?

File: game/src/plugins/cmd/src/skill-cmd.plugin.kts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                skills.addExperience(skill, SkillSet.MAXIMUM_EXP)
+            }
+        }
+
+/**
+ * Levels the specified skill to the specified level, optionally updating the current level as well.
+ */
+on_command("level", PrivilegeLevel.ADMINISTRATOR)
+        .then { player ->
+            val invalidSyntax = "Invalid syntax - ::level [skill-id] [level]"
```
Comment (Major-) [inline game/src/plugins/cmd/src/skill-cmd.plugin.kts]: Can this message be updated to account for the third arg

File: game/src/plugins/cmd/src/skill-cmd.plugin.kts (hunk 6 above)
Comment (cubeee) [inline game/src/plugins/cmd/src/skill-cmd.plugin.kts]: Addressed in all commands: f561f25

File: game/src/plugins/cmd/src/skill-cmd.plugin.kts (hunk 7 above)
Comment (cubeee) [inline game/src/plugins/cmd/src/skill-cmd.plugin.kts]: Done

**Comment by cubeee** [comment]
@Major- all but ::lookup converted, ready for merge when there is no more improvements to make

File: game/src/plugins/cmd/src/item-cmd.plugin.kts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+            val id = arguments[0].toInt()
+            val amount = if (arguments.size == 2) arguments[1].toInt() else 1
+
+            if (id < 0 || id >= ItemDefinition.count()) {
+                player.sendMessage("The item id you specified is out of bounds!")
+                return@then
+            }
+
+            if (amount < 0 || amount > Int.MAX_VALUE) {
```
Comment (Major-) [inline game/src/plugins/cmd/src/item-cmd.plugin.kts]: Yeah this is originally from the ruby plugin, because the backing type for an integer is swapped out if it overflows. Don't have that problem here

File: game/src/plugins/cmd/src/messaging-cmd.plugin.kts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import org.apollo.game.model.World
+import org.apollo.game.model.entity.setting.PrivilegeLevel
+
+var world: World? = null
+
+start { world ->
```
Comment (cubeee) [inline game/src/plugins/cmd/src/messaging-cmd.plugin.kts]: Done for all plugins

File: game/src/plugins/cmd/src/messaging-cmd.plugin.kts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    this.world = world
+}
+
+on_command("broadcast", PrivilegeLevel.ADMINISTRATOR)
+        .then { player ->
+            val message = arguments.joinToString(" ")
+            val broadcast = "[Broadcast] ${player.username.capitalize()}: $message"
+
+            world?.playerRepository?.forEach { other ->
```
Comment (cubeee) [inline game/src/plugins/cmd/src/messaging-cmd.plugin.kts]: Done for all plugins

File: game/src/plugins/cmd/src/item-cmd.plugin.kts (hunk 8 above)
Comment (cubeee) [inline game/src/plugins/cmd/src/item-cmd.plugin.kts]: Done, 9108e88
